src/Data.Json.JTable.Internal.purs

```diff
@@ -0,0 +1,214 @@
+                   k' = snoc k nt
+               in T p w' h' k'
+
+-- produce a tree of header data from json
+tFromJson :: JPath -> Json -> Tree
+tFromJson path json = let
+  obj = toObject json <#> \jo ->
+    if M.isEmpty jo then T path 1 0 []    
+    else let k = M.toList jo <#> uncurry \l j -> tFromJson (snoc path l) j
+             w = k <#> tWidth # foldl (+) 0
+             h = 1 + foldl max 0 (k <#> tHeight)
+         in T path w h k
+  m = (>>=) (toArray json)
+  tuple = m \a -> (widthOfPrimTuple path a) <#> \n -> T path n 0 []
+  array = m \a -> let
+    ts = a <#> tFromJson path
+    t' = foldl tMergeArray (T path 0 0 []) (ts >>= tKids)
+    w' = t' # tWidth
+    mw = ((head $ nub (ts <#> tWidth)) <#> max w') <|> (Just $ max 1 w')
+    in mw <#> \w -> T path w (t' # tHeight) (t' # tKids) 
+  prim = T path 1 0 []
+  in prim `fromMaybe` (obj <|> tuple <|> array)
+
+
+-- merge table segments for each key of an object into one
+cMergeObj :: [(Tuple Number Table)] -> Table
+cMergeObj rss = let
+  maxh = rss <#> snd <#> length # foldl max 0
+  in (0 .. maxh-1) <#> \n -> rss >>= uncurry \w rs ->
+    if length rs == 1 then if n == 0 
+                           then AU.head rs <#> \(C c w h j) -> C c w maxh j
+                           else [] `fromMaybe` (rs !! n)
+    else [C (JCursorTop) w 1 primNull] `fromMaybe` (rs !! n)
+
+-- maybe merge a tuple of objects into a table segment
+mergeObjTuple ::Tree -> JCursor -> [Json] -> Maybe Table
+mergeObjTuple t@(T p w h k) c ja = head ja *> do
+  jos <- for ja toObject
+  let keyss = jos <#> M.keys
+  let all_keys = concat keyss
+  so ((length all_keys) == (length $ nub all_keys)) $ 
+    cMergeObj $ k <#> \(t'@(T p' w' _ _)) -> let 
+      label = AU.last p'
+      i = findIndex (elem label) keyss
+      j = jnull `fromMaybe` (jos !! i >>= M.lookup label)
+      in Tuple w' (cFromJson t' (downField label $ downIndex i c) j)
+
+-- produce data table from json, according to header tree
+cFromJson :: Tree -> JCursor -> Json -> Table
+cFromJson t@(T p w h k) c json = let
+  obj = toObject json <#> \jo ->
+    if M.isEmpty jo then [[C c w 1 primNull]] 
+    else cMergeObj $ do
+      t'@(T p' w' _ _) <- k
+      let label = AU.last p'
+      let j = jnull `fromMaybe` M.lookup label jo
+      return $ Tuple w' (cFromJson t' (downField label c) j)
+  ma = toArray json
+  obtup = ma >>= mergeObjTuple t c
+  width = ma >>= widthOfPrimTuple p >>= so (h <= 0 && w > 1)
+  tuple = ma <#> \a -> singleton $ (0 .. w-1) <#> \i ->
+    C (downIndex i c) 1 1 $ toPrim $ jnull `fromMaybe` (a !! i)
+  array = ma <#> \a -> zipWithIndex a >>= uncurry
+                  \j i -> cFromJson t (downIndex i c) j
+  prim = [[C c w 1 $ toPrim json]]
+  in prim `fromMaybe` (obj <|> obtup <|> (width *> tuple) <|> array)
+
+
+-- render a grid from an array of arrays
+renderRows :: forall a. (Markup -> Markup) -> (Number -> Number -> a -> Markup) -> [[a]] -> Markup
+renderRows tr' cellf rows = mconcat $ do
```

Comment:
This is a good example of keeping row/col implicit "until the last possible moment", which means this code can have no errors about row/col.